lib/puppet/util/character_encoding.rb

```diff
@@ -62,13 +63,27 @@ def convert_to_utf_8!(string)
           # (it is valid in its current encoding), we can guess this string was
           # not originally unicode. Transcode it to UTF-8. For strings with
           # original encodings like SHIFT_JIS, this should be the final result.
+          # If the string comes to us as BINARY encoded, we don't know what it
+          # started as. However, to encode! we need a starting place, and our
+          # best guess (only guess) is whatever the system currently is
+          # (default_external). So set external_encoding to default_external
+          # before we try to transcode to UTF-8. There's one exception, which is
+          # if external encoding is currently UTF-8. We don't bother
+          # force_encoding to UTF-8 first because: a) once we do, ruby won't
+          # transcode it for us to UTF-8 anymore and b) we've already previously
+          # determined that this string is not valid UTF-8 bytes so force
+          # encoding to UTF-8 isn't going to help us
+          string.force_encoding(Encoding.default_external) if string.encoding == Encoding::BINARY && Encoding.default_external != Encoding::UTF_8
           return string.encode!(Encoding::UTF_8)
         else
           # If the string is neither valid UTF-8 as-is nor valid in its current
           # encoding, fail. It requires user remediation.
           raise EncodingError
         end
       rescue EncodingError => detail
+        # Ensure the string retains it original external encoding since
+        # we've failed
+        string.force_encoding(original_encoding)
```

Comment:
Ah that makes sense, thanks. Yes I agree with your concern. In its first form the util never modified the original string and always returned a copy. The performance implications of always copying the string (in some cases multiple times) were not insignificant, so we adjusted to modifying the string in place. But, as you point out, we ended up with inconsistent return semantics. I'm inclined to modify the string in place and always return nil to keep the performance improvement but improve consistency.